src/java/org/apache/cassandra/index/sai/disk/v1/bitpack/AbstractBlockPackedReader.java

```diff
@@ -43,6 +44,7 @@ public abstract class AbstractBlockPackedReader implements LongArray
         this.valueCount = valueCount;
         this.input = new SeekingRandomAccessInput(indexInput);
         this.blockBitsPerValue = blockBitsPerValue;
+        this.readers = new LongValues[blockBitsPerValue.length];
```

Comment:
How many will there be typically?

Reply:
It's the number of blocks. It looks like it is governed by:

```java
    public static int numBlocks(long size, int blockSize) {
        int numBlocks = (int)(size / (long)blockSize) + (size % (long)blockSize == 0L ? 0 : 1);
        if ((long)numBlocks * (long)blockSize < size) {
            throw new IllegalArgumentException("size is too large for this block size");
        } else {
            return numBlocks;
        }
    }
```

We already create several other arrays of that length:

```java
        final int numBlocks = numBlocks(valueCount, meta.blockSize);
        blockBitsPerValue = new byte[numBlocks];
        blockOffsets = new long[numBlocks];
        minValues = new long[numBlocks];
```

We could use a map to keep memory utilization low in the case where we don't need to access all of the blocks.

Reply:
Based on looking at the usage, it probably makes sense to use a map here since the usage seems to vary quite a bit in terms of how many readers we actually need, and it seems pretty reasonable that we won't need many in several cases.